aws_signing_helper/pkcs11_signer.go

```diff
@@ -0,0 +1,214 @@
+
+	if err = p.FindObjectsInit(session, templatecrt); err != nil {
+		fmt.Println("nil, err")
+	}
+	defer func() {
+		finalErr := p.FindObjectsFinal(session)
+		if err == nil {
+			err = finalErr
+		}
+	}()
+
+	objs, b, err := p.FindObjects(session, 1)
+	for err == nil {
+		var o []pkcs11.ObjectHandle
+		o, b, err = p.FindObjects(session, 1)
+		if err != nil {
+			continue
+		}
+		if len(o) == 0 {
+			break
+		}
+		objs = append(objs, o...)
+	}
+	if err != nil {
+		log.Println("Failed to find: ", b)
+		if len(objs) == 0 {
+			fmt.Println("nil")
+		}
+	}
+
+	_ = p.FindObjectsFinal(session)  // found nothing
+
+	for i := range objs {
+		attributescrt := []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_VALUE, 0),
+		}
+
+		if attributescrt, err = p.GetAttributeValue(session, objs[i], attributescrt); err != nil {
+			fmt.Println("nil, err2")
+		}
+
+		rawCertificate := attributescrt[0].Value
+		cert, err := x509.ParseCertificate(rawCertificate)
+		if err != nil {fmt.Println("Error parsing certificate")}
+		serialNumber := cert.SerialNumber.String()
+		certLocated.Subject = cert.Subject.CommonName 
+		certLocated.Issuer = cert.Issuer.CommonName
+		certLocated.SerialNumber, _ = new(big.Int).SetString(serialNumber, 16)
+		if reflect.DeepEqual(certLocated, certIdentifier) {
+			return i, cert, nil
+		} else {
+		}
+	}
+
+
+	return 1, nil, errors.New("unsupported certificate")
+}
+
+func (pkcs11Signer PKCS11Signer) Public() crypto.PublicKey {
+	return nil
+}
+
+func (pkcs11Signer PKCS11Signer) Close() {
+}
+
+func (signer PKCS11Signer) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+		lib := signer.libpkcs11
+		pin := signer.pinpkcs11
+		id := signer.idpkcs11
+
+		p := pkcs11.New(lib) 
+		err = p.Initialize()
+		if err != nil {
+			panic(err)
+		}
+
+		defer p.Destroy()
+		defer p.Finalize()
+
+		slots, err := p.GetSlotList(true)
+		if err != nil {
+			panic(err)
+		}
+
+		session, err := p.OpenSession(slots[0], pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+		if err != nil {
+			panic(err)
+		}
+		defer p.CloseSession(session)
+
+		err = p.Login(session, pkcs11.CKU_USER, pin)// err = p.Login(session, pkcs11.CKU_USER, pin) 
+		if err != nil {
+			panic(err)
+		}
+		defer p.Logout(session)
+
+		templatepk := []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+		}
+		_ = p.FindObjectsInit(session, templatepk) 
+		objspk, _, _ := p.FindObjects(session, 100)
+		_ = p.FindObjectsFinal(session)  // found nothing
+
+		err = p.Login(session, pkcs11.CKU_CONTEXT_SPECIFIC, pin)
+
+		err = p.SignInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_SHA256_RSA_PKCS, nil)}, objspk[id]) // CKM_SHA256_RSA_PKCS
+		if err != nil {
+		fmt.Println("err: ", err)
+			log.Fatalf("Signing Initiation failed (%s)\n", err.Error())
+		}
+
+		sig, err := p.Sign(session, digest)
+		if err != nil {
+		err = fmt.Errorf("Signing failed (%s)\n", err.Error())
+		fmt.Println("err: ", err)
```

Comment:
Minor, but the message here should be more descriptive.

Reply:
The Signing failed from the Sign call provides the details on current issue if I am correct.